# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [6]:
import pandas as pd
url='https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv'

df1=pd.read_csv(url)
df=df1.copy()
display(df.head(2))
df.info()

,Unnamed: 0,Customer,State,Customer Lifetime Value,Response,Coverage,Education,Effective To Date,EmploymentStatus,Gender,...,Number of Open Complaints,Number of Policies,Policy Type,Policy,Renew Offer Type,Sales Channel,Total Claim Amount,Vehicle Class,Vehicle Size,Vehicle Type
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2/18/11,Employed,M,...,0.0,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,NaN
1,1,KX64629,California,2228.525238,No,Basic,College,1/18/11,Unemployed,F,...,0.0,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,NaN


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     10910 non-null  int64  
 1   Customer                       10910 non-null  object 
 2   State                          10279 non-null  object 
 3   Customer Lifetime Value        10910 non-null  float64
 4   Response                       10279 non-null  object 
 5   Coverage                       10910 non-null  object 
 6   Education                      10910 non-null  object 
 7   Effective To Date              10910 non-null  object 
 8   EmploymentStatus               10910 non-null  object 
 9   Gender                         10910 non-null  object 
 10  Income                         10910 non-null  int64  
 11  Location Code                  10910 non-null  object 
 12  Marital Status                 10910 non-null 

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

In [7]:
#DataFrame based on total_claim_amount<1000 and response Yes to marketing campaign
print("null values for total claim amount?")
display(df['Total Claim Amount'].isnull().sum())
print("null values for Response?")
display(df['Response'].isnull().sum())
#Now we can create the DataFrame based on total_claim_amount<1000 and response Yes 
claim_df=df[(df['Response']=='Yes') & (df['Total Claim Amount']<1000) ]
display(claim_df.shape)
claim_df.head(5)



null values for total claim amount?


np.int64(0)

null values for Response?


np.int64(631)

(1399, 26)

,Unnamed: 0,Customer,State,Customer Lifetime Value,Response,Coverage,Education,Effective To Date,EmploymentStatus,Gender,...,Number of Open Complaints,Number of Policies,Policy Type,Policy,Renew Offer Type,Sales Channel,Total Claim Amount,Vehicle Class,Vehicle Size,Vehicle Type
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,1/11/11,Employed,M,...,0.0,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A
8,8,FM55990,California,5989.773931,Yes,Premium,College,1/19/11,Employed,M,...,0.0,1,Personal Auto,Personal L1,Offer2,Branch,739.200000,Sports Car,Medsize,NaN
15,15,CW49887,California,4626.801093,Yes,Basic,Master,1/16/11,Employed,F,...,0.0,1,Special Auto,Special L1,Offer2,Branch,547.200000,SUV,Medsize,NaN
19,19,NJ54277,California,3746.751625,Yes,Extended,College,2/26/11,Employed,F,...,1.0,1,Personal Auto,Personal L2,Offer2,Call Center,19.575683,Two-Door Car,Large,A
27,27,MQ68407,Oregon,4376.363592,Yes,Premium,Bachelor,2/28/11,Employed,F,...,0.0,1,Personal Auto,Personal L3,Offer2,Agent,60.036683,Four-Door Car,Medsize,NaN


2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

In [3]:
#to see the average monthly_premium and/or customer lifetime value by policy_type and gender for 
#customers who responded "Yes", first we need to check how those columns are formated and their values.

In [4]:
display(df['Policy Type'].unique())
display(df['Gender'].unique())

array(['Corporate Auto', 'Personal Auto', 'Special Auto'], dtype=object)

array(['M', 'F'], dtype=object)

In [5]:
#Now we can create a pivot table to see the average monthly_premium and/or customer lifetime value by policy_type and gender for 
#customers who responded "Yes"
col=['Monthly Premium Auto','Customer Lifetime Value']
yes_piv= df[df.Response=='Yes'].pivot_table(index=['Policy Type','Gender'], values=col,aggfunc='mean')
display(yes_piv.round(2))


Customer Lifetime Value  Monthly Premium Auto
Policy Type    Gender                                               
Corporate Auto F                       7712.63                 94.30
               M                       7944.47                 92.19
Personal Auto  F                       8339.79                 99.00
               M                       7448.38                 91.09
Special Auto   F                       7691.58                 92.31
               M                       8247.09                 86.34

In [6]:
#compare these insights to total_claim_amount patterns

In [7]:
#comparing to the DataFrame based on total_claim_amount<1000 and response Yes to marketing campaign
claim_df.head(2)
claim_piv= claim_df.pivot_table(index=['Policy Type','Gender'], values=col,aggfunc='mean')
display(claim_piv.round(2))


Customer Lifetime Value  Monthly Premium Auto
Policy Type    Gender                                               
Corporate Auto F                       7334.77                 89.04
               M                       7920.40                 88.55
Personal Auto  F                       7966.93                 90.62
               M                       7481.82                 87.43
Special Auto   F                       7594.01                 86.71
               M                       8348.23                 80.00

In [18]:
#comparing to the DataFrame based on total_claim_amount<1000 and response Yes to marketing campaign
pd.concat([yes_piv.round(2),claim_piv.round(2)], axis=1)

Customer Lifetime Value  Monthly Premium Auto  \
Policy Type    Gender                                                  
Corporate Auto F                       7712.63                 94.30   
               M                       7944.47                 92.19   
Personal Auto  F                       8339.79                 99.00   
               M                       7448.38                 91.09   
Special Auto   F                       7691.58                 92.31   
               M                       8247.09                 86.34   

                       Customer Lifetime Value  Monthly Premium Auto  
Policy Type    Gender                                                 
Corporate Auto F                       7334.77                 89.04  
               M                       7920.40                 88.55  
Personal Auto  F                       7966.93                 90.62  
               M                       7481.82                 87.43  
Special Auto   F                       7594.01                 86.71  
               M                       8348.23                 80.00

Interpreting results:

Looking at the groups that responded Yes to the campaign, without a claim amount threshold, the females with the policy Personal Auto are the ones who brind the higher CLV (8339.79 USD). When considering the threshold of claim amount (<1000 USD), it shows that the highest CLV comes from the males with policy type Special Auto (8348.23).

If we look at the CLV with and without the claim amount threshold (<1000 USD), then we can say that the highest CLV comes from the Males who had low claim amount and with Special Auto Policy Type who bring 8348.23 USD of CLV in average, followed by the Females with policy type Personal Auto who bring 8339.79 USD CLV in average. These would be then the most profitable segments.



3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [8]:
#Analyze the total number of customers who have policies in each state
#as there are nan values for State, we will not consider those
cust_piv=df[df.State.isnull()==False].pivot_table(index='State',values='Customer',aggfunc='count',fill_value=0,margins=True, margins_name='Total')
cust_piv

,Customer
State,
Arizona,1937
California,3552
Nevada,993
Oregon,2909
Washington,888
Total,10279


In [20]:
#from the table above we can see that all states have above 500 customer, being California with the highest
#number of customer, with 3552.

In [21]:
#the number of customer who have more than 500 customers are
cust_piv[cust_piv.Customer>500]

,Customer
State,
Arizona,1937
California,3552
Nevada,993
Oregon,2909
Washington,888
Total,10279


In [19]:
#Another way to do it as AI reviewer suggests
state_counts=df.groupby('State')['Customer'].count()
display(state_counts)
result=state_counts[state_counts >500]
result

State
Arizona       1937
California    3552
Nevada         993
Oregon        2909
Washington     888
Name: Customer, dtype: int64

State
Arizona       1937
California    3552
Nevada         993
Oregon        2909
Washington     888
Name: Customer, dtype: int64

In [ ]:
#from the table above we can see that all states have above 500 customer, being California with the highest
#number of customer, with 3552.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [22]:
#Getting the customer lifetime value by education level and gender and agg by(maximum, minimum, and median )
df.groupby(['Education', 'Gender'])['Customer Lifetime Value'].agg(['max', 'min', 'median']).round(2)

max      min   median
Education            Gender                            
Bachelor             F       73225.96  1904.00  5640.51
                     M       67907.27  1898.01  5548.03
College              F       61850.19  1898.68  5623.61
                     M       61134.68  1918.12  6005.85
Doctor               F       44856.11  2395.57  5332.46
                     M       32677.34  2267.60  5577.67
High School or Below F       55277.45  2144.92  6039.55
                     M       83325.38  1940.98  6286.73
Master               F       51016.07  2417.78  5729.86
                     M       50568.26  2272.31  5579.10

In [22]:
# Observations:
#- The maximum CLV comes from the Men with High School or Below education with CLV = 83325.38 USD and while the maximum from women comes from the ones with the ones with Bachelor with a CLV of 73225.96 USD.
#- The minimum CLV comes from males with Bachelor (1898.01 USD) which is comparable to the women with collegue (1898.68), just cents difference.
#- The highest median comes from the males with High School or Below, with CLV 6286.73 USD and the lowes median comes from Females wit Doctor degree, with CLV of 5332.46 USD.

# Interpretation: 
#CLV does not scale monotonically with education. Education level alone doesn't predict customer value — lower-education groups actually contain the highest-spending outliers, and gender effects depend on which education group you're looking at.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here